src/coreclr/jit/importer.cpp

```diff
@@ -10462,7 +10464,8 @@ void Compiler::impImportBlockCode(BasicBlock* block)
                                 GenTree* boxPayloadAddress =
                                     gtNewOperNode(GT_ADD, TYP_BYREF, cloneOperand, boxPayloadOffset);
                                 GenTree* nullcheck = gtNewNullCheck(op1, block);
-                                GenTree* result    = gtNewOperNode(GT_COMMA, TYP_BYREF, nullcheck, boxPayloadAddress);
+                                nullcheck->gtFlags |= GTF_ORDER_SIDEEFF;
+                                GenTree* result = gtNewOperNode(GT_COMMA, TYP_BYREF, nullcheck, boxPayloadAddress);
```

Comment:
It is not obvious what the ordering effect here protects against -- what would a theoretical example of bad transformation look like?

Reply:
I think I can revert this one too given that `boxPayloadAddress` is known to be very small when `cloneOperand` is null.
If it wasn't then early prop would probably be able to do something illegal here if it removed this null check due to an upcoming null check, causing us to form some random byref.

Reply:
Hmm, I'm not sure after all. Without setting this then if we did the optimization in #71435 we could potentially remove these null checks and turn a NRE into an AV when accessing a field right around the explicit null check threshold. Consider something like:
```scala
ASG
  LCL_VAR byref V00
  COMMA byref           // added here
    NULLCHECK            // node A
      LCL_VAR ref V01
    ADD byref
      LCL_VAR ref V01
      CNS_INT long 8

ASG
  LCL_VAR int V02
  IND int               // node B
    ADD
      LCL_VAR byref V00
      CNS_INT long 0xFFFC

ASG
  LCL_VAR int V03
  IND int             // node C
    LCL_VAR ref V01
```

The optimization would reason as follows: the node A null check is unnecessary because node B only has the same side effects as the null check (throwing NRE), and node C subsumes node A. However, node B will actually AV without node A.

Reply:
I would classify this as the optimization being wrong, since we can only consider NRE to be "true" NRE and not "NRE but maybe AV" with small offsets (we do the checks against the max null check offset in other places to differentiate these two cases, though not everywhere).

More fundamentally, fixing this with an ordering effect seems... arbitrary? What is the ordering dependency?

Reply:
We've talked with @jakobbotsch on Discord and concluded the ordering here is needed because "almost null" byrefs must never occur on dynamically reachable paths, so swapping the `ADD` with the null check would effectively introduce UB into the program.

Reply:
> It has also been determined that our handling around `GTF_ORDER_SIDEEFF` is a bit lacking, and it's meaning a bit unclear. We've settled on the "only fully side-effect-less nodes can be reordered with ordered nodes"

It might be useful to add the meaning as a comment. Maybe next to the definition in gentree.h? The current "sub-expression has a re-ordering side effect" is a little lacking.

Reply:
I agree we should have some more detailed docs on what the effect flags mean and how they are assumed to be set, but I think it would be better to add a section in https://github.com/dotnet/runtime/blob/main/docs/design/coreclr/jit/ryujit-tutorial.md and refer to that. I will aim to do a write-up of that soon.

Reply:
ok, I'm happy with that.